File: examples/backend-go/analytics.go
package main

import (
	"context"
	"log/slog"

	voidhash "github.com/voidhashcom/voidhash-go"
)

// analyticsClient records product analytics through the SDK.
//
// Two different credentials are in play, which is the thing worth noticing:
//
//   - Capture posts to event ingest, which authenticates on the publishable
//     key (voidhash.WithPublishableKey). It is disabled when that is unset.
//   - SetAttributes is a server-to-server write on the secret key. Traits
//     describe the person and persist, so facts like the current plan go here
//     rather than being repeated on every event.
type analyticsClient struct {
	client  *voidhash.Client
	enabled bool
	logger  *slog.Logger
}

func newAnalyticsClient(client *voidhash.Client, publishableKey string, logger *slog.Logger) *analyticsClient {
	return &analyticsClient{client: client, enabled: publishableKey != "", logger: logger}
}

// Enabled reports whether events can be sent at all.
func (a *analyticsClient) Enabled() bool {
	return a.enabled
}

// Capture sends one event.
func (a *analyticsClient) Capture(ctx context.Context, distinctID, event string, properties map[string]any) error {
	_, err := a.client.EventCapture.Capture(ctx, voidhash.Event{
		DistinctID: distinctID,
		Event:      event,
		Properties: properties,
	})
	return err
}

// SetAttributes writes person traits, creating the person when the distinct id
// is new. Unlike Capture this needs no publishable key, so it works even when
// analytics capture is switched off.
func (a *analyticsClient) SetAttributes(ctx context.Context, distinctID string, traits map[string]any) error {
	_, err := a.client.Persons.SetAttributes(ctx, voidhash.SetPersonAttributesParams{
		DistinctID: distinctID,
		Traits:     traits,
	})
	return err
}

File: examples/backend-php/src/Nimbus/Analytics.php
<?php

declare(strict_types=1);

namespace Voidhash\Example\Nimbus;

use Voidhash\Example\Logger;
use Voidhash\Exception\ApiException;
use Voidhash\Generated\Core\Model\SetPersonAttributesBodyJsonEncoding;
use Voidhash\VoidhashClient;

/**
 * Server-side analytics, over the SDK.
 *
 * Two different credentials are in play, which is the thing worth noticing:
 *
 * - {@see Analytics::capture()} posts to event ingest, which authenticates on
 *   the **publishable** key (`publishableKey` in
 *   {@see VoidhashClient::create()}). It is a no-op when that is unset.
 * - {@see Analytics::setAttributes()} is a server-to-server write on the
 *   **secret** key. Traits describe the person and persist, so facts like the
 *   current plan go here rather than being repeated on every event.
 */
final class Analytics
{
    public const NOTE_CREATED = 'note_created';
    public const EXPORT_REQUESTED = 'export_requested';
    public const PAYWALL_VIEWED = 'paywall_viewed';
    public const CHECKOUT_STARTED = 'checkout_started';

    public function __construct(
        private readonly VoidhashClient $client,
        private readonly Logger $logger,
    ) {
    }

    /** Whether a publishable key is configured; when false every capture is a no-op. */
    public function isEnabled(): bool
    {
        return $this->client->eventCapture->isEnabled();
    }

    /**
     * Captures an event, logging and swallowing any failure.
     *
     * Use this for events the product emits as a side effect. Nobody should
     * lose a note because the analytics pipeline had a bad minute.
     *
     * @param array<string, mixed> $properties
     */
    public function capture(string $event, string $distinctId, array $properties = []): void
    {
        try {
            $this->captureOrFail($event, $distinctId, $properties);
        } catch (ApiException $exception) {
            $this->logger->warning('event capture failed', [
                'event' => $event,
                'distinctId' => $distinctId,
                'cause' => $exception->getMessage(),
            ]);
        }
    }

    /**
     * Captures an event and lets failures propagate. Returns false when
     * analytics is not configured, which is not a failure.
     *
     * Use this when forwarding an event is the whole point of the request, as
     * on `POST /v1/events`: the caller asked for one thing, so tell them the
     * truth about whether it happened.
     *
     * @param array<string, mixed> $properties
     *
     * @throws ApiException when ingest rejected the event
     */
    public function captureOrFail(string $event, string $distinctId, array $properties = []): bool
    {
        if (!$this->isEnabled()) {
            return false;
        }

        $this->client->eventCapture->capture([
            'event' => $event,
            'distinctId' => $distinctId,
            'properties' => $properties,
        ]);

        return true;
    }

    /**
     * Writes person traits, logging and swallowing any failure. Unlike capture
     * this uses the secret key, so it works even when analytics is switched off.
     *
     * @param array<string, string|int|float|bool|null> $traits
     */
    public function setAttributes(string $distinctId, array $traits): void
    {
        try {
            $this->client->persons->setAttributes(
                (new SetPersonAttributesBodyJsonEncoding())
                    ->setDistinctId($distinctId)
                    ->setTraits($traits),
            );
        } catch (ApiException $exception) {
            $this->logger->warning('writing person attributes failed', [
                'distinctId' => $distinctId,
                'cause' => $exception->getMessage(),
            ]);
        }
    }
}
